### PR TITLE
p&f e2e: the test should log expected and actual response header in case it fails. 

### PR DESCRIPTION
**Pull request by tkashem**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR adds logging for P&F e2e, the test logs expected and actual response header in case it fails. 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

```docs

```

### CODE REVIEW

File: test/e2e/apimachinery/flowcontrol.go [hunk 1]
```diff
@@ -59,13 +59,14 @@ var _ = SIGDescribe("API priority and fairness", func() {
...
 
-		ginkgo.By("checking response headers contain flow-schema/priority-level uid")
-		if !testResponseHeaderMatches(f, matchingUsername, string(createdPriorityLevel.UID), string(createdFlowSchema.UID)) {
-			framework.Failf("matching user doesnt received UID for the testing priority-level and flow-schema")
-		}
-		if testResponseHeaderMatches(f, nonMatchingUsername, string(createdPriorityLevel.UID), string(createdPriorityLevel.UID)) {
-			framework.Failf("non-matching user unexpectedly received UID for the testing priority-level and flow-schema")
-		}
+		var response *http.Response
+		ginkgo.By("checking response headers contain flow-schema/priority-level uid for a matching user")
```
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:63]: How about we write "and" instead of "/", and capitalize UID, in these phrases?

**Review by MikeSpreitzer — APPROVED** [review]
This looks functional enough, but the wording could be improved a little.

File: test/e2e/apimachinery/flowcontrol.go [hunk 2]
```diff
@@ -51,21 +51,26 @@ var _ = SIGDescribe("API priority and fairness", func() {
...
-		ginkgo.By("creating a testing flowschema")
+		ginkgo.By("creating a testing flow-schema")
 		createdFlowSchema, cleanup := createFlowSchema(f, testingFlowSchemaName, 1000, testingPriorityLevelName, matchingUsername)
 		defer cleanup()
 
-		ginkgo.By("checking response headers contain flow-schema/priority-level uid")
-		if !testResponseHeaderMatches(f, matchingUsername, string(createdPriorityLevel.UID), string(createdFlowSchema.UID)) {
-			framework.Failf("matching user doesnt received UID for the testing priority-level and flow-schema")
-		}
-		if testResponseHeaderMatches(f, nonMatchingUsername, string(createdPriorityLevel.UID), string(createdPriorityLevel.UID)) {
```
Comment (tkashem) [inline test/e2e/apimachinery/flowcontrol.go:66]: @MikeSpreitzer it seems we have been passing `createdPriorityLevel.UID` for both flow-schema and priority-level. If I am not mistaken, for a non matching user, we expect the UID of catch-all objects in the response header.

Below, I am validating that a non matching user gets a catch-all UID in the response header. Let me know if this looks good.

**Review by yue9944882 — COMMENTED** [review]
/lgtm

File: test/e2e/apimachinery/flowcontrol.go [hunk 3]
```diff
@@ -334,15 +351,16 @@ func makeRequest(f *framework.Framework, username string) *http.Response {
 	return response
 }
 
-func testResponseHeaderMatches(f *framework.Framework, impersonatingUser, plUID, fsUID string) bool {
-	response := makeRequest(f, impersonatingUser)
-	if response.Header.Get(flowcontrol.ResponseHeaderMatchedFlowSchemaUID) != fsUID {
-		return false
+func verifyResponseHeaderMatch(f *framework.Framework, response *http.Response, plUIDWant, fsUIDWant string) {
```
Comment (adtac) [inline test/e2e/apimachinery/flowcontrol.go:354]:
```suggestion
func verifyResponseHeaderMatch(response *http.Response, plUIDWant, fsUIDWant string) {
```

don't need the first argument anymore

**Comment by adtac** [comment]
Related to https://github.com/kubernetes/kubernetes/issues/96803 btw

File: test/e2e/apimachinery/flowcontrol.go [hunk 4]
```diff
@@ -51,21 +51,26 @@ var _ = SIGDescribe("API priority and fairness", func() {
 		matchingUsername := "noxu"
 		nonMatchingUsername := "foo"
 
-		ginkgo.By("creating a testing prioritylevel")
+		ginkgo.By("creating a testing priority-level")
```
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:54]: I think we usually write simply "priority level", and either "flow schema" or "FlowSchema".

File: test/e2e/apimachinery/flowcontrol.go [hunk 5]
```diff
@@ -51,21 +51,26 @@ var _ = SIGDescribe("API priority and fairness", func() {
...
+		globalDefaultPriorityLevel := getPriorityLevelConfig(f, globalDefaultFlowSchema.Spec.PriorityLevelConfiguration.Name)
+
+		var response *http.Response
+		ginkgo.By("checking response headers contain flow-schema and priority-level UID for a matching user")
+		response = makeRequest(f, matchingUsername)
+		verifyResponseHeaderMatch(response, string(createdPriorityLevel.UID), string(createdFlowSchema.UID))
+
+		ginkgo.By("checking response headers contain catch-all flow-schema and priority-level UID for a non-matching user")
+		response = makeRequest(f, nonMatchingUsername)
+		verifyResponseHeaderMatch(response, string(globalDefaultPriorityLevel.UID), string(globalDefaultFlowSchema.UID))
```
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:73]: I do not recommend expecting any particular classification for a non-matching request.  That is a matter of configuation that this test is not controlling.  We could test that the relevant headers are present but should not insist on them identifying any particular flow schema or priority level.

Comment (tkashem) [reply test/e2e/apimachinery/flowcontrol.go:73]: sounds good to me.

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm

File: test/e2e/apimachinery/flowcontrol.go [hunk 6]
```diff
@@ -45,26 +45,37 @@ const (
...
+		if plUIDWant, plUIDGot := string(createdPriorityLevel.UID), getPriorityLevelUID(response); plUIDWant != plUIDGot {
+			framework.Failf("expected PriorityLevelConfiguration UID in the response header: %s, but got: %s, response header: %#v", plUIDWant, plUIDGot, response.Header)
 		}
-		if testResponseHeaderMatches(f, nonMatchingUsername, string(createdPriorityLevel.UID), string(createdPriorityLevel.UID)) {
-			framework.Failf("non-matching user unexpectedly received UID for the testing priority-level and flow-schema")
+		if fsUIDWant, fsUIDGot := string(createdFlowSchema.UID), getFlowSchemaUID(response); fsUIDWant != fsUIDGot {
+			framework.Failf("expected FlowSchema UID in the response header: %s, but got: %s, response header: %#v", fsUIDWant, fsUIDGot, response.Header)
+		}
+
+		ginkgo.By("response headers should contain UID of the default FlowSchema and PriorityLevelConfiguration for a non-matching user")
```
Comment (adtac) [inline test/e2e/apimachinery/flowcontrol.go:72]: I believe this description needs to be updated slightly as well

**Comment by tkashem** [comment]
`pull-kubernetes-e2e-gce-ubuntu-containerd` seems to be flaky - https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/96966/pull-kubernetes-e2e-gce-ubuntu-containerd/1334143128373825536/.

It has failed a few times for this PR.

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm

**Comment by MikeSpreitzer** [comment]
`pull-kubernetes-e2e-gce-ubuntu-containerd` failed for https://github.com/kubernetes/kubernetes/pull/96985 too

**Comment by MikeSpreitzer** [comment]
#97002

**Comment by MikeSpreitzer** [comment]
The pull-kubernetes-e2e-gce-ubuntu-containerd job is still failing everything.

**Pull request merged**